### ISSUE TITLE
256 Color Support in ConEmu on Windows

### ISSUE THREAD

**Issue by dancread** [vifm/vifm#426]
Hello,
I am wondering if 256 color support has ever been tested in ConEmu. There seems to be support for 256 colors and I do have 256 colors working in vim and a few other applications on ConEmu:
https://conemu.github.io/en/AnsiEscapeCodes.html.

As a quick hack, I disabled the native Linux terminal check and was able to get 256 Color previews in bat vs just plain text output in the latest release for cmder (which uses ConEmu as a console emulator).
![vifm_bat_256](https://user-images.githubusercontent.com/6148998/56470809-4a480900-6410-11e9-8a39-6684308a0045.PNG)
![vifm_bat_default](https://user-images.githubusercontent.com/6148998/56470810-4ae09f80-6410-11e9-9857-936ba55e4995.PNG)

[win32_256_color.patch.txt](https://github.com/vifm/vifm/files/3101231/win32_256_color.patch.txt)

**Comment by xaizek** [comment]
Hi,

I'm not sure your changes could cause that. Those checks are related to attributes instead of colors and to file highlighting instead of preview. Windows API for drawing in console seems to have only [4 bits](https://msdn.microsoft.com/en-us/windows/ms682013(v=vs.80)) per foreground and background. 

Those applications seem to be using escape sequences, but version of pdcurses that vifm uses doesn't do that. However, development was continued recently and it looks like there is a support [now](https://github.com/wmcbrine/PDCurses/blob/master/wincon/pdcdisp.c#L140-L168). Switching to newer release needs to be investigated as there are vifm-specific patches which need to be ported (although some fixes were applied by the author, I think).

**Comment by dancread** [comment]
Yes, you are right. I just reverted my changes and recompiled the master branch and I am getting extra colors without any error messages. I was trying to track this back by enabling a 256 colorscheme on windows. I found the error message and traced the calls down to the cmd_handler and just removed anything that would block the colorscheme change. I see now that probably had nothing to do with it. I think what is actually making this work is that I am linking against PDCurses 3.8 with WIDE=Y. It looks like that would enable what your link is revering to. I did read through the vifm curses patches and saw some of them were implemented in the main PDCurses branch and some were not (namely, the ALT_PERIOD patch which was needed for compilation). Either way, I was hoping to demonstrate 256 colors can be done relatively easily in a "native" fashion without a proper terminal emulator on windows. Perhaps this is something that can be implemented? I can help, but I obviously do not know the code very well as evident by my hack.

**Comment by dancread** [comment]
I just looked through the patches.  I can port them to PDCurses 3.8 and see what happens. I had to hack up the Makefile.win a bit to compile things with mingw, so I am not sure how reliable my build is for testing.

**Comment by xaizek** [comment]
> I am linking against PDCurses 3.8 with WIDE=Y

That explains why it suddenly started to work.

Since you managed to build vifm against 3.8, it looks like porting should be manageable.

> I had to hack up the Makefile.win a bit to compile things with mingw

I'm using MinGW as well, although possibly of a different version.

> I can port them to PDCurses 3.8 and see what happens.

That would be nice! Since PDCurses is back in development, updated patches could also be sent upstream later (I stopped doing this when old bug tracker and forum got disabled).

**Comment by dancread** [comment]
I went through the patches by hand and reapplied them where I could to versions 3.8. Some rewrites of methods address the same problems (I hope). There were two patches that have not been addressed in PDCurses 3.8 and I have attached the new versions to be applied to the 3.8 tag of PDCurses and a new exe.

[pdcurses-1-window-timeout.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-1-window-timeout.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/e50f65fa92ba69e7204b99a80105f8b515fd45bb
[pdcurses-2-win32-alt-dot-key.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-2-win32-alt-dot-key.patch) -> pdcurses_1-win32-alt-dot-key.patch
[pdcurses-3-fix-underline.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-3-fix-underline.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/18464b052777b3ff3b8a5fc92c452d8ed3eb17a6
[pdcurses-4-fix-reverse.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-4-fix-reverse.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/18464b052777b3ff3b8a5fc92c452d8ed3eb17a6
[pdcurses-5-fix-cjk-chars-display.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-5-fix-cjk-chars-display.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/18464b052777b3ff3b8a5fc92c452d8ed3eb17a6
[pdcurses-6-resize-detection-and-manual-handling.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-6-resize-detection-and-manual-handling.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/f52f0e354a16da7df8eea49efa00834adf0b4b59
[pdcurses-7-win32-is_termresized.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-7-win32-is_termresized.patch) -> pdcurses-2-win32-is_termresized.patch
[pdcurses-8-win32-mark-resize-key-as-key-code.patch](https://github.com/vifm/vifm/blob/2ea0d52bc2afb119727dc0d5aa249a6151338531/patches/pdcurses/pdcurses-8-win32-mark-resize-key-as-key-code.patch) -> Addressed in https://github.com/wmcbrine/PDCurses/commit/f52f0e354a16da7df8eea49efa00834adf0b4b59

**Comment by dancread** [comment]
[vifm_pdcurses_3.8.zip](https://github.com/vifm/vifm/files/3102021/vifm_pdcurses_3.8.zip)

**Comment by dancread** [comment]
I am only testing with https://github.com/sharkdp/bat and the results look good. I did switch to a 256 color scheme which vifm allowed me to do, but there are some very odd screen artifacts. Doubling lines, disappearing text, etc. Keeping to my original  16 color scheme works great though. 256 color schemes probably need to be disallowed if this patch is used.

**Comment by xaizek** [comment]
Thank you. I'll try it out, will try to see where the artifacts come from (if they show up in color schemes, they might as well appear in preview as well) and then go through the changes to see if everything works as expected.

**Comment by xaizek** [comment]
> [...] there are some very odd screen artifacts. Doubling lines, disappearing text, etc.

Could you check if this gets better on `bug-fixes` branch? I saw some weird stuff happening even with 16 color scheme on v0.10 and fixed it. As a test I also tried to force pdcurses to use ansi codes exclusively, which caused other artifacts. So either of the projects might be to blame or both of them. With the fix both 256 color scheme and preview seem to work fine for me in ConEmu.

**Comment by dancread** [comment]
Yes, I can try later today.

**Comment by dancread** [comment]
I just downloaded the latest version of cmder mini to test the version built from the bug-fixes branch and it seems to be much better. There were still some visual artifacts like blocks of colors not updating with screen refreshes and the border decorations being different colors (see the attached screen shot, the red blocks in the preview are left over from the OtherLine color attribute). The is much better than the garbled text I was seeing before. I am testing with juef-zenburn since that seems to be a problematic colorscheme. I also noticed some cases where the active pane cursor showed the value of the file above it, but I could not re create it with any consistency. 
![zenburn-256](https://user-images.githubusercontent.com/6148998/57583775-94655d00-7499-11e9-9a5f-30fde8537dd7.png)

**Comment by xaizek** [comment]
Great. Thanks. With `juef-zenburn` I see weird stuff and I think that [this code](https://github.com/wmcbrine/PDCurses/blob/7c00f33999426568743420224d5e5fd84314e75a/wincon/pdcdisp.c#L112-L122) in PDCurses has something to do with it. It's working around line wrapping in some way and I see screen being scrolled up as if console window does this automatically.

I might try to see if I can understand what's going on. If this is the only issue with new PDCurses, then support of 256 color schemes can be advertised as experimental (`near-default` doesn't have any issues, while `juef-zenburn` is really broken for me).

**Comment by xaizek** [comment]
It must be this undesired scrolling, I figured out that `juef-zenburn` works fine if you comment out setting highlighting for `CmdLine` or avoid using 256 colors for it. So if you want to use it, just do that.

I tried various things in PDCurses, but screen still gets messed up and I believe this is related to https://github.com/Maximus5/ConEmu/issues/1444, screenshots there appear very similar to what I see. So it should be fine to leave as is, because it's an issue of ConEmu, which might not even have a reliable workaround for PDCurses (I managed to keep text intact, but highlighting still gets scrolled by ConEmu).
